src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -1581,6 +1581,62 @@ public DynamicObject replace(DynamicObject array, DynamicObject other,
 
     }
 
+    @Primitive(name = "array_rotate_inplace", needsSelf = false, lowerFixnum = 2)
+    @ImportStatic(ArrayGuards.class)
+    public abstract static class RotateInplaceNode extends PrimitiveArrayArgumentsNode {
+
+        @Specialization(guards = "strategy.matches(array)", limit = "ARRAY_STRATEGIES")
+        public DynamicObject rotate(DynamicObject array, int amount,
+                @Cached("of(array)") ArrayStrategy strategy) {
+            final int size = strategy.getSize(array);
+            assert 0 < amount && amount < size;
+            final ArrayMirror mirror = strategy.newMirror(array);
+
+            rotateReverse(amount, size, mirror);
+
+            return array;
+        }
+
+        protected void rotateReverse(int amount, int size, ArrayMirror mirror) {
+            final int last = size - 1;
+            reverse(mirror, amount, last);
+            reverse(mirror, 0, amount - 1);
+            reverse(mirror, 0, last);
+        }
+
+        private void reverse(ArrayMirror mirror, int from, int to) {
+            while (from < to) {
+                final Object tmp = mirror.get(from);
+                mirror.set(from, mirror.get(to));
+                mirror.set(to, tmp);
+                from++;
+                to--;
+            }
+        }
+
+        protected void rotateElementByElement(int amount, int size, ArrayMirror mirror) {
```

Comment:
I'm not seeing where this is called from.

Reply:
That's an alternative implementation, I need to measure which one is better.
The reversing one does `2n` reads+writes, while this one only `n` reads+writes but the reversing implementation OTOH accesses memory in a contiguous fashion not jumping around like `rotateElementByElement`.